aten/src/ATen/core/jit_type.h

```diff
@@ -1123,6 +1126,88 @@ struct CAFFE2_API TupleType : public NamedType {
   std::shared_ptr<FunctionSchema> schema_;
 };
 
+struct EnumType;
+using EnumTypePtr = std::shared_ptr<EnumType>;
+struct CAFFE2_API EnumType : public NamedType {
+  friend struct Type;
+  static const TypeKind Kind = TypeKind::EnumType;
+
+  static EnumTypePtr create(const c10::QualifiedName& qualified_name, TypePtr value) {
+    switch (value->kind()) {
+      case TypeKind::AnyType:
+      case TypeKind::IntType:
+      case TypeKind::FloatType:
+      case TypeKind::StringType:
+      case TypeKind::TensorType:
+        return EnumTypePtr(new EnumType(qualified_name, value));
+      default:
+        AT_ERROR(
+            "Cannot create Enum with value type '",
+            value->str(),
+            "', only int, float, Tensor and string keys are supported");
+    }
+  }
+
+  std::string str() const override {
+    return "Enum<" + annotation_str() + ">";
+  }
+
+  std::string repr_str() const override {
+    return str();
+  }
+
+  TypePtr getValueType() const {
```

Comment:
I believe we're also going to need to store a list of (enum_name, value) for each item in the enum, although I know this is 1/N so understood if it's NYI...